graph/cycle_detection.cpp

```diff
@@ -0,0 +1,115 @@
+#include <vector> //path saving and node colouring
```

Comment:
Please describe how the algorithm works, a Wikipedia or any other source for algorithm explanation, what the algorithm does, etc..